### PR TITLE
Create TC build to upload Gutenberg source-maps to Sentry [PoC/WIP]

### PR DESCRIPTION
**Pull request by fullofcaffeine**
Project: p9oQ9f-18L-p2

#### Changes proposed in this Pull Request

Create a dedicated TC build to upload source-maps for a specific GB version to Sentry. For now, it will use the fixed dummy Sentry release `wpcom-test-01` (which we probably should rename, actually). It will clean up all artifacts for this release and re-upload the new ones, each time.

The idea is for this build to be triggered from our installation script. We should only proceed to the actual activation after the source-maps have been uploaded (so we'll need to do some polling in the script to see when the build has finished). 

This implementation is not ideal because it uses a single fixed Sentry release (`wpcom-test-01`). This can cause false-positives due to race conditions when releases:

1. Source-maps finished uploading but the version in WPCOM is still old, there's a window of time when errros can happen and the new source-maps will be used
2. When we activate Sentry for AT, this will be the case most of the time, as there's always a delay for updating Gutenberg on AT.

The scenarios above are not ideal but having source-maps is better than having none. That being said...

### Food for thought for a follow-up:

The ideal implementation will use a new Sentry release for each deployment. The problem is that Sentry is not exclusive to Gutenberg, and its instance shared across the WPAdmin context, which means that a release for the WPAdmin project can't have metadata exclusive to a certain Gutenberg release. We can't create a new Sentry project for Gutenberg either, as it needs to be specified in the Sentry configuration in the frontend, and it's shared across the execution spae for GB, ETK, et al.

The best we can do, I think, is to just have sequential releases, like each release of Gutenberg (which will trigger this build) and other WPCOM plugins just create a new sequential release on Sentry. 

Another possibility is to look at the WPCOM git repo as the consolidator of all these apps that are tracked by the Sentry WPadmin project in the frontend. We can somehow get the commit hash for a deployment and use it to create the new release in Sentry. This has another race-condition downside though: we can only get the hash (AFAIK) after the deployment is done, so in theory, there's going to be a window of time where the release will not have source-maps. Also, we should remember that for each of those releases, the hash should be committed together so that the Sentry configuration in `error-reporting` [can access it and pass it to Sentry](https://github.com/Automattic/wp-calypso/blob/trunk/apps/editing-toolkit/editing-toolkit-plugin/error-reporting/index.js#L21) (i.e window.WPCOM_RELEASE or something like that)

### TODO
- [ ] Test the build and:
- [ ] Add the SENTRY_AUTH_TOKEN to the WPComTests project, if not already there
- [ ] How handle failures/success in this build? It should only fail if one of the commands in the bash node script fail, is it already handled by TC?
- [ ] Should the build be part of the WPComPlugins project?  Should we move it somewhere else? Semantically, in part, I think it does, but it doesn't follow the same build pattern as the others, though.
- [ ] If / once it tests well, modify the installation script to trigger this build before deploying.

#### Testing instructions

TBD

Related to #

### CODE REVIEW

File: .teamcity/_self/projects/WPComPlugins.kt [hunk 1]
```diff
@@ -184,3 +188,54 @@ private object InlineHelp : WPComPluginBuild(
...
+			)
+		}
+
+		steps {
+			mergeTrunk()
+			bashNodeScript {
+				name = "Upload source maps to Sentry"
+				scriptContent = """
+					# Install the Sentry CLI binary
+					curl -sL https://sentry.io/get-cli/ | bash
```
Comment (noahtallen) [inline .teamcity/_self/projects/WPComPlugins.kt:210]: We might be able to put this in one of our docker images instead of running it every time. For example, the wpcom image has a bunch of stuff installed in it: https://github.com/Automattic/wp-calypso/blob/29b2eb19eb032b1dbad22395e1cdb91611ce225c/Dockerfile.base#L128

File: .teamcity/_self/projects/WPComPlugins.kt [hunk 2]
```diff
@@ -184,3 +188,54 @@ private object InlineHelp : WPComPluginBuild(
...
+
+					wget https://github.com/WordPress/gutenberg/releases/tag/%GUTENBERG_VERSION%/gutenberg.zip
+					unzip gutenberg.zip -d gutenberg
+					cd gutenberg
+
+					sentry-cli --auth-token %SENTRY_AUTH_TOKEN% releases --org a8c files "wpcom-test-01" delete --all
```
Comment (noahtallen) [inline .teamcity/_self/projects/WPComPlugins.kt]: I'm guessing we'll need to pass the release via a paramater?

So we'd basically create some kind of release identifier on wpcom, update it in the sentry ETK code (?) and then also make sure it's used here?

Part of the challenge will be keeping the version updated on wpcom without too much overhead.

File: .teamcity/_self/projects/WPComPlugins.kt [hunk 3]
```diff
@@ -184,3 +188,54 @@ private object InlineHelp : WPComPluginBuild(
...
+					cd apps/editing-toolkit/editing-toolkit-plugin
```
Comment (noahtallen) [inline .teamcity/_self/projects/WPComPlugins.kt]: Is this needed?

**Comment by noahtallen** [comment]
> This has another race-condition downside though: we can only get the hash (AFAIK) after the deployment is done, so in theory, there's going to be a window of time where the release will not have source-maps. Also, we should remember that for each of those releases, the hash should be committed together so that the Sentry configuration in error-reporting [can access it and pass it to Sentry](https://github.com/Automattic/wp-calypso/blob/trunk/apps/editing-toolkit/editing-toolkit-plugin/error-reporting/index.js#L21) (i.e window.WPCOM_RELEASE or something like that)

Yeah, super good points.

The sequential release ID makes sense, knowing that we need to generate the release before deployment. In theory, there's a gap between commit and deploy, but it's so small compared to Calypso.

Atomic is also challenge, as we would need to do a full ETK release to simply update the release constant on all platforms. That's a lot of extra work. 

Perhaps we need to expose the release ID another way? For example, we created the Calypso apps cache buster JSON file to sort of help with this same problem. We could possibly use that as well:
- update release in the JSON file
- on the remote site, download the JSON file and cache it for, say, 15min
- use the version ID from the JSON file for the sentry version.

A benefit is that it's really easy to update the JSON file using `jq` in one of our install script, and we have a helper function for interacting with it on wpcom via PHP already.

But that does seem fairly complex. 🤔 

The other challenge is ensuring that _all_ sourcemaps are present in a release. Did we determine that Sentry will _not_ consider sourcemaps that were uploaded as part of a previous release? So do we have to upload all sourcemaps for all plugins anytime we create a release with the CLI?

File: .teamcity/_self/projects/WPComPlugins.kt (hunk 3 above)
Comment (fullofcaffeine) [reply .teamcity/_self/projects/WPComPlugins.kt]: Opsie, nop! Copy-pasta mistake :)

File: .teamcity/_self/projects/WPComPlugins.kt (hunk 2 above)
Comment (fullofcaffeine) [reply .teamcity/_self/projects/WPComPlugins.kt]: Yes, exactly. The release is being passed there as "wpcom-test-01". I think it's fine to leave it as "wpcom-test-01" for now, but I tried to elaborate on this problem in the PR description, quoting here: 


> The ideal implementation will use a new Sentry release for each deployment. The problem is that Sentry is not exclusive to Gutenberg, and its instance shared across the WPAdmin context in the frontend, which means that a release for the WPAdmin project can't have metadata exclusive to a certain Gutenberg release. We can't create a new Sentry project for Gutenberg either, as it needs to be specified in the Sentry configuration in the frontend, and it's shared across the execution spae for GB, ETK, et al.
> 
> The best we can do, I think, is to just have sequential releases, like each release of Gutenberg (which will trigger this build) and other WPCOM plugins just create a new sequential release on Sentry.
> 
> Another possibility is to look at the WPCOM git repo as the consolidator of all these apps that are tracked by the Sentry WPadmin project in the frontend. We can somehow get the commit hash for a deployment and use it to create the new release in Sentry. This has another race-condition downside though: we can only get the hash (AFAIK) after the deployment is done, so in theory, there's going to be a window of time where the release will not have source-maps. Also, we should remember that for each of those releases, the hash should be committed together so that the Sentry configuration in error-reporting [can access it and pass it to Sentry](https://github.com/Automattic/wp-calypso/blob/trunk/apps/editing-toolkit/editing-toolkit-plugin/error-reporting/index.js#L21) (i.e window.WPCOM_RELEASE or something like that)

**Comment by fullofcaffeine** [comment]
> The other challenge is ensuring that all sourcemaps are present in a release. Did we determine that Sentry will not consider sourcemaps that were uploaded as part of a previous release?

Yep, I've contacted support about this, they confirmed that each release starts fresh in terms of artifacts.

**Comment by fullofcaffeine** [comment]
> Atomic is also challenge, as we would need to do a full ETK release to simply update the release constant on all platforms. That's a lot of extra work.
> Perhaps we need to expose the release ID another way?
> A benefit is that it's really easy to update the JSON file using jq in one of our install script, and we have a helper function for interacting with it on wpcom via PHP already.

I think this is a very good idea! I totally agree that it'd be best to avoid attaching an ETK release everytime we deploy to simple sites. 

There are other nuances, like for example, a simple release (the Sentry artifact) might not semantically justify a new Sentry release for AT (unless we're updating something that affects AT, too). I think it'd be best to have its own Sentry instance and project, or at the very least, we should pass an AT (or `siteType`) metadata somewhere so that it's clear the issue was generated from an AT site. But we can think about those when we start the project to get Sentry on AT.

**Comment by noahtallen** [comment]
Ah, that's a good point. We could even just have the release on Atomic be like `atomic-wp-admin-$version` and then simple be like `simple-wp-admin-$version`.

Or it could be a different "environment." I believe sentry has logic to handle different releases in different environments

File: .teamcity/_self/projects/WPComPlugins.kt (hunk 1 above)
Comment (fullofcaffeine) [reply .teamcity/_self/projects/WPComPlugins.kt:210]: How often are those images built? 

I think this is a good idea, but I'll probably do it as a follow-up. I want to confirm it works well first :)

File: .teamcity/_self/projects/WPComPlugins.kt [hunk 4]
```diff
@@ -184,3 +188,39 @@ private object InlineHelp : WPComPluginBuild(
...
+				name = "GUTENBERG_VERSION",
+				value = "",
+				label = "Gutenberg version",
+				description = "The Gutenberg version to upload source maps for (include the whole string, including the `v` prefix)",
+				allowEmpty = false
+			)
+		}
+
+		steps {
+			mergeTrunk()
```
Comment (fullofcaffeine) [inline .teamcity/_self/projects/WPComPlugins.kt]: Oh, I don't think this will be needed for this build :)

File: .teamcity/_self/projects/WPComPlugins.kt [hunk 5]
```diff
@@ -184,3 +188,38 @@ private object InlineHelp : WPComPluginBuild(
 	docsLink = "TODO",
 	withPRNotify = "false",
 )
+
+private object GutenbergUploadSourceMapsToSentry: BuildType() {
```
Comment (noahtallen) [inline .teamcity/_self/projects/WPComPlugins.kt:190]: Nit: we could (probably?) change this syntax to:

```
... : BuildType({
  name = ""
  params {
  ...
})
```

Basically removing the need to call `init` directly

File: .teamcity/_self/projects/WPComPlugins.kt [hunk 6]
```diff
@@ -1,8 +1,11 @@
 package _self.projects
 
 import _self.bashNodeScript
+import _self.lib.utils.mergeTrunk
```
Comment (noahtallen) [inline .teamcity/_self/projects/WPComPlugins.kt]: Not needed anymore

File: .teamcity/_self/projects/WPComPlugins.kt [hunk 7]
```diff
@@ -184,3 +188,38 @@ private object InlineHelp : WPComPluginBuild(
...
+		name = "Upload Gutenberg Source Maps to Sentry";
+		id("WPComPlugins_GutenbergUploadSourceMapsToSentry");
+
+		params {
+			text(
+				name = "GUTENBERG_VERSION",
+				value = "",
+				label = "Gutenberg version",
+				description = "The Gutenberg version to upload source maps for (include the whole string, including the `v` prefix)",
+				allowEmpty = false
```
Comment (noahtallen) [inline .teamcity/_self/projects/WPComPlugins.kt:201]: Nice, I didn't know about this syntax :)

File: .teamcity/_self/projects/WPComPlugins.kt (hunk 1 above)
Comment (noahtallen) [reply .teamcity/_self/projects/WPComPlugins.kt:210]: the base image is rebuilt once a day, but we can always trigger the build manually to get a fresh copy

File: .teamcity/_self/projects/WPComPlugins.kt [hunk 8]
```diff
@@ -184,3 +188,38 @@ private object InlineHelp : WPComPluginBuild(
...
+				name = "Upload source maps to Sentry"
+				scriptContent = """
+					# Install the Sentry CLI binary
+					curl -sL https://sentry.io/get-cli/ | bash
+
+					wget https://github.com/WordPress/gutenberg/releases/tag/%GUTENBERG_VERSION%/gutenberg.zip
+					unzip gutenberg.zip -d gutenberg
+					cd gutenberg
+
+					sentry-cli --auth-token %SENTRY_AUTH_TOKEN% releases --org a8c files "wpcom-test-01" delete --all
```
Comment (noahtallen) [inline .teamcity/_self/projects/WPComPlugins.kt]: What's the purpose of deleting the existing files?

Comment (fullofcaffeine) [reply .teamcity/_self/projects/WPComPlugins.kt]: For now, we only have a single, fixed release, and I thought I'd start with a clean slate at every deploy, which will make it more manageable in the long run. The release is only used for GB at the moment, so that doesn't have any further consequences. I think, though, that we might as well not delete anything, since the paths will be different anyway for each upload (due to the version change), and if we use this release for other app's source-maps, then deleting here would be indeed a problem.

TL;DR: I think we can get rid of it.

Comment (noahtallen) [reply .teamcity/_self/projects/WPComPlugins.kt]: Cool! That's similar to what I was thinking, especially if we need to upload other apps

Comment (fullofcaffeine) [reply .teamcity/_self/projects/WPComPlugins.kt]: Another benefit of not deleting is that we can have source-maps for multiple releases at the same time 🤔 which can be a good thing for enabling them on edge before we push to production.

File: .teamcity/_self/projects/WPComPlugins.kt (hunk 6 above)
Comment (fullofcaffeine) [reply .teamcity/_self/projects/WPComPlugins.kt]: Ah yeah!

**Review by noahtallen — APPROVED** [review]
Looks like a good starting point!

File: .teamcity/_self/projects/WPComPlugins.kt (hunk 5 above)
Comment (fullofcaffeine) [reply .teamcity/_self/projects/WPComPlugins.kt:190]: Thanks, took note and will look in a follow-up :)
